### PR TITLE
Fix verifyCommits error when push a new branch

### LINKED ISSUE #25565: Pushing new branch fails if new branch is covered by rule requiring signed commits
### Description

If a new branch is pushed, and the repository has a rule that would require signed commits for the new branch, the commit is rejected with a 500 error regardless of whether it's signed.

When pushing a new branch, the "old" commit is the empty ID (0000000000000000000000000000000000000000).  verifyCommits has no provision for this and passes an invalid commit range to git rev-list.  Prior to 1.19 this wasn't an issue because only pre-existing individual branches could be protected.

I was able to reproduce with https://try.gitea.io/CraigTest/test, which is set up with a blanket rule to require commits on all branches.

### Gitea Version

1.19.3

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

https://gist.github.com/Craig-Holmquist-NTI/1267366c5beb28eb38b8c49cd3de7f5c

### Screenshots

_No response_

### Git Version

2.41.0.windows.1

### Operating System

Windows 10

### How are you running Gitea?

gitea-1.19.3-windows-4.0-amd64.exe binary downloaded from gitea.com, run from command line in a simple test environment

### Database

SQLite

### PR DESCRIPTION
**Pull request by CaiCandong**
> ### Description
> If a new branch is pushed, and the repository has a rule that would require signed commits for the new branch, the commit is rejected with a 500 error regardless of whether it's signed.
> 
> When pushing a new branch, the "old" commit is the empty ID (0000000000000000000000000000000000000000). verifyCommits has no provision for this and passes an invalid commit range to git rev-list. Prior to 1.19 this wasn't an issue because only pre-existing individual branches could be protected.
> 
> I was able to reproduce with [try.gitea.io/CraigTest/test](https://try.gitea.io/CraigTest/test), which is set up with a blanket rule to require commits on all branches.

Fix #25565
Very thanks to @Craig-Holmquist-NTI for reporting the bug and suggesting an valid solution!

### CODE REVIEW

**Comment by lunny** [comment]
Could you add some test?

File: routers/private/hook_verification.go [hunk 1]
```diff
@@ -28,23 +28,28 @@ func verifyCommits(oldCommitID, newCommitID string, repo *git.Repository, env []
 		_ = stdoutWriter.Close()
 	}()
 
+	var command *git.Command
+	if oldCommitID == git.EmptySHA {
+		command = git.NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(newCommitID).AddArguments("--not", "--all")
```
Comment (wxiaoguang) [inline routers/private/hook_verification.go:36]:
```suggestion
		// When creating a new branch, the oldCommitID is empty, by using "newCommitID --not --all":
		// List commits that are reachable by following the newCommitID, exclude "all" existing heads/tags commits
		// So, it only lists the new commits received, doesn't list the commits already present in the receiving repository
		command = git.NewCommand(repo.Ctx, "rev-list").AddDynamicArguments(newCommitID).AddArguments("--not", "--all")
```

At least, it needs comments

Comment (CaiCandong) [reply routers/private/hook_verification.go:36]: thinks

**Comment by lunny** [comment]
> > Could you add some test?
> 
> how to add some test for it ?

You can create a unit test for the function `verifyCommits`.

**Comment by CaiCandong** [comment]
I don't know how to prepare the test environment in the unit test, anyone help me to write the test?

**Comment by wxiaoguang** [comment]
Merge history is broken

**Comment by lunny** [comment]
> Merge history is broken

Fixed

**Comment by CaiCandong** [comment]
This unit test requires the appropriate user data, PGP signature commit, and PGP  Key to complete, which will take a bit of time to complete. Maybe we can merge this PR first (I've tested it locally with no problems) and I'll open a PR to supplement this test.@lunny

File: routers/private/hook_verification_test.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+	"github.com/stretchr/testify/assert"
+)
+
+var testReposDir = "tests/repos/"
+
+func TestVerifyCommits(t *testing.T) {
+	unittest.PrepareTestEnv(t)
+
+	gitRepo, err := git.OpenRepository(context.Background(), testReposDir+"repo1_hook_verification")
```
Comment (lunny) [inline routers/private/hook_verification_test.go:21]: defer gitRepo.Close()

Comment (CaiCandong) [reply routers/private/hook_verification_test.go:21]: Done https://github.com/go-gitea/gitea/pull/26664/commits/a37097b461e45692f075f9194c8e920a3daa9d80

File: models/fixtures/email_address.yml [hunk 3]
```diff
@@ -276,4 +276,12 @@
...
   is_activated: false
-  is_primary: false
\ No newline at end of file
+  is_primary: false
+
+-
+  id: 36
+  uid: 36
+  email: 1290147055@qq.com
+  lower_email: 1290147055@qq.com
```
Comment (wxiaoguang) [inline models/fixtures/email_address.yml]: Please do not use personal information

File: models/fixtures/user.yml [hunk 4]
```diff
@@ -1301,7 +1301,7 @@
   lower_name: limited_org36
   name: limited_org36
   full_name: Limited Org 36
-  email: limited_org36@example.com
+  email: 1290147055@qq.com
```
Comment (wxiaoguang) [inline models/fixtures/user.yml]: Please do not do this.

File: models/fixtures/gpg_key.yml [hunk 5]
```diff
@@ -1 +1,23 @@
-[] # empty
+-
+  id: 5
+  owner_id: 36
+  key_id: 2ECAAA351B755A40
+  primary_key_id:
+  content: xsDNBGTjB1cBDACv7YEaiA5VaoTBR0fcPRy+RHgTd88ALhklVVd54lbMKWwsHa3Ms9sCYbq8chly4g50l45o8qL0mu+zkZDvrU3EsZAwFwv/6iKltQXRIQ7j0uNKIyJi/+yyMiRJtU24ptAkc+bHhb2wi7cnJtldGyDZIoymzotTdpJupmjVFXqQtbeqMHpSKg6JdyGLK+wX+ITD7gZ6UddPPbw5Yo6N5b56w55cSIG6cDtafo+/Fc4WF0jdDDmCEyPlr3afDKEcpmnDSJsMhm7/eZ5MRn2dh9+brnEFyxkKSJYfI7ptA6trAs+DCil9FgZbXM9P/c0joVUzi1lSGSu1ryav9eAb4J5lgzl2Tad338+cVWuCHR6yTbq4qZcfgNVnFK1WMUrg9MF2A8wNLO1O9LSQpXwDSBmnM9W7TcTE6bpxbADVH7WEScCNmExKeqUnvvypaeiwmwz7Jqh/V5NhjE3ySgo3j/oBFFUp2SSkZhk7p0uP8mDLSdfPLxtBUv3EQAPs8CCD7WkAEQEAAQ==
```
Comment (wxiaoguang) [inline models/fixtures/gpg_key.yml]: Please provide related private key to make future maintenance easier.

File: models/fixtures/email_address.yml (hunk 3 above)
Comment (silverwind) [reply models/fixtures/email_address.yml]: I don't see this too strict, but if it's easily fixed, do it.

File: models/fixtures/gpg_key.yml (hunk 5 above)
Comment (silverwind) [reply models/fixtures/gpg_key.yml]: Best to create a dummy keypair and commit both public and private, if we not already have one.

File: models/fixtures/email_address.yml (hunk 3 above)
Comment (CaiCandong) [reply models/fixtures/email_address.yml]: Done

File: models/fixtures/user.yml (hunk 4 above)
Comment (CaiCandong) [reply models/fixtures/user.yml]: Done

File: models/fixtures/gpg_key.yml (hunk 5 above)
Comment (CaiCandong) [reply models/fixtures/gpg_key.yml]: Done [`dca58bf` (#26664)](https://github.com/go-gitea/gitea/pull/26664/commits/dca58bfcdc34f2c9033950152741b35d69ca0c64)

**Comment by lunny** [comment]
A small nit, some git files could be removed like `routers/private/tests/repos/repo1_hook_verification/description` and `routers/private/tests/repos/repo1_hook_verification/exclude`. Otherwise LGTM

**Review by lunny — APPROVED** [review]

**Comment by CaiCandong** [comment]
> A small nit, some git files could be removed like `routers/private/tests/repos/repo1_hook_verification/description` and `routers/private/tests/repos/repo1_hook_verification/exclude`. Otherwise LGTM

Done [`dc24b94` (#26664)](https://github.com/go-gitea/gitea/pull/26664/commits/dc24b94bcb4eb5409b11927f7ce5950da0a322b0)

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**